crates/derive/src/pipeline/pipeline_test.rs

```diff
@@ -0,0 +1,378 @@
+use prost::Message;
+use protocol::flow::{
+    derive_api::{self, Code, Config, DocHeader},
+    CollectionSpec, DerivationSpec, LambdaSpec, TransformSpec,
+};
+use serde_json::{json, Value};
+use tempfile::TempDir;
+use tuple::{TupleDepth, TuplePack};
+use url::Url;
+
+#[test]
+fn test_pipeline_stats() {
+    let mut fixture = FixtureBuilder::new("test-derive-pipeline")
+        .key(&["/key"])
+        .schema(json!({
+            "type": "object",
+            "properties": {
+                "key": {"type": "string"},
+                "sum": {"type": "integer", "reduce": {"strategy": "sum"}},
+            },
+            "reduce": {"strategy": "merge"}
+        }))
+        .add_transform()
+        // Transform 0 should have both update and publish stats.
+        // The outputs are doubled here so we can verify on the output stats.
+        .update(|val| Ok(vec![val.clone(), val]))
+        .publish(|_, _, reg| Ok(vec![reg.clone(), reg]))
+        .end_transform()
+        .add_transform()
+        // The second transform is just a publish-only identity function
+        .publish(|source, _, _| Ok(vec![source]))
+        .end_transform()
+        .add_transform()
+        // The third transform only updates, no publish
+        .update(|val| Ok(vec![val]))
+        .end_transform()
+        .finish();
+
+    fixture.add_source_document(0, key("bee"), r#"{"key": "bee", "sum": 2}"#.as_bytes());
+    fixture.add_source_document(0, key("bee"), r#"{"key": "bee", "sum": 1}"#.as_bytes());
+    fixture.add_source_document(1, key("knee"), r#"{"key": "knee", "sum": 7}"#.as_bytes());
+    fixture.add_source_document(1, key("bee"), r#"{"key": "bee", "sum": 66}"#.as_bytes());
+
+    // These should only show up as 2 registers created
+    fixture.add_source_document(2, key("tea"), r#"{"key": "tea", "sum": 9}"#.as_bytes());
+    fixture.add_source_document(2, key("tea"), r#"{"key": "tea", "sum": 1}"#.as_bytes());
+    fixture.add_source_document(2, key("free"), r#"{"key": "free", "sum": 42}"#.as_bytes());
+
+    let (docs, stats) = fixture.poll_to_completion();
+    // Redact the time output on the snapshot so that it's deterministic.
+    insta::assert_yaml_snapshot!(stats, {
+        ".**.total_seconds" => "time-redacted",
+    });
+    assert_eq!(2, docs.len());
+}
+
+fn key(k: impl TuplePack) -> Vec<u8> {
+    let mut out = Vec::new();
+    k.pack(&mut out, TupleDepth::new().increment()).unwrap();
+    out
+}
+
+/// A helper for testing a `Pipeline` by rosoving trampoline tasks (lambdas) using Rust functions.
+struct Fixture {
+    _temp_dir: TempDir,
+    pipeline: Pipeline,
+    transforms: Vec<TransformFixture>,
+}
+
+#[derive(Default)]
+struct TransformFixture {
+    update: Option<Box<dyn FnMut(Value) -> Result<Vec<Value>, ()>>>,
+    /// publish lambda parameters are `(source, prev_register, new_register)`
+    publish: Option<Box<dyn FnMut(Value, Value, Value) -> Result<Vec<Value>, ()>>>,
+}
+
+impl Fixture {
```

Comment:
This is great, thanks for adding it

Reply:
Thanks! I started feeling a little uncertain about it after seeing how big it got, but hopefully it'll be useful in more tests or benchmarks in the future.